README.md
# Hyperion-Lite

[![License GPL 3][badge-license]][COPYING]
[![wercker status](https://app.wercker.com/status/866edc2dd3b4ed63731381b426e93a2c/s "wercker status")](https://app.wercker.com/project/bykey/866edc2dd3b4ed63731381b426e93a2c)

## Description

[Hyperion-Lite][] is a [Docker][] (>= 0.11) image (Ubuntu 14.04 based) containing :
* [Hyperion-Lite][] web description : `http://xxx:9990`
* [Elasticsearch][] (v1.3.2) web interface : `http://xxx:9992/elasticsearch/`
* [Grafana][] (v1.7.0) web interface : `http://xxx:9990/grafana/`
* [Kibana][] (v3.1.0) web interface : `http://xxx:9990/kibana/`
* [Graphite][] (v0.10.0) web interface : `http://xxx:9990/graphite/`
* [Statsd][] (v0.7.1) daemon on `8125` and `8126`
* [Supervisor][] is used to manage processes.
* [InfluxDB][] (v0.8.0) web interface : `http://xxx:8083`

Some [Elasticsearch][] plugins are available:
* [ElasticSearchHead][]: `http://xxx:9992/_plugin/head/`
* [ElasticHQ][]: `http://xxx:9992/_plugin/HQ/`
* [Kopf][]: `http://xxx:9992/_plugin/kopf/`


![Hyperion-Lite dashboard](hyperion-lite.png)


## Deployment

### Local

Use this [script](client/hyperion.sh) to help you to local deployment :
```bash
$ client/hyperion.sh help
Usage: client/hyperion.sh <command>
Commands:
  pull      :    Pull the Hyperion-Lite image from the registry
  start     :    Start Hyperion-Lite container
  stop      :    Stop Hyperion-Lite container
  help      :    Display this help
```

* Install it:

        $ ./hyperion.sh pull && ./hyperion.sh start

* For [Graphite][] metrics, you could test your local installation using [hyperion_statsd.py](client/hyperion_statsd.py):

        $ pip install -r requirements.txt
        $ ./client/hyperion_statsd.py

* For [InfluxDB][] metrics, you could test your installation using [sysinfo_influxdb][]:

        $ curl -X POST 'http://localhost:8086/db?u=root&p=root' \
            -d '{"name": "hyperion-lite"}'
        $ sysinfo_influxdb -host 127.0.0.1:8086 -P hyperion-lite -d hyperion-lite -v=text -D

* Go to `http://localhost:9990/grafana` to see dashboard metrics


### Virtualbox

A `Vagrantfile` using [CoreOS][] (version 367.1.0) is provided if you want to use it in a virtual machine.

* Install dependencies : [Virtualbox][] (>= 4.3.10), [Vagrant][] (>= 1.6)

* Launch VM:

        $ vagrant up

* Test your [Graphite][] metrics using [hyperion_client.py](client/hyperion_statsd.py):

        $ ./hyperion_statsd.py -s 10.2.4.6 -p 8125

* Test your [InfluxDB][] metrics using [sysinfo_influxdb][]:

        $ curl -X POST 'http://10.2.4.6:8086/db?u=root&p=root' \
            -d '{"name": "hyperion-lite"}'
        $ sysinfo_influxdb -host 10.2.4.6:8086 -P hyperion-lite -d hyperion-lite -v=text -D

* Go to `http://10.2.4.6:9990/grafana/`

* You could connect to your virtual machine by ssh to manage your installation
using [CoreOS][] tools ([Etcd][] and [Fleet][] and [Systemd][]).

        $ vagrant ssh
        fleetctl list-units
        UNIT			STATE		LOAD	ACTIVE	SUB	DESC		MACHINE
        hyperion-lite.service	launched	loaded	active	running	hyperion-lite	43824b21.../10.2.4.6
        $ fleetctl status hyperion-lite.service
        ● hyperion-lite.service - hyperion-lite
           Loaded: loaded (/etc/systemd/system/hyperion-lite.service; linked-runtime)
           Active: active (running) since Tue 2014-07-07 23:28:58 UTC; 15min ago
           Process: 3620 ExecStartPost=/usr/bin/etcdctl set /hyperion-lite/host ${COREOS_PUBLIC_IPV4} (code=exited, status=0/SUCCESS)
           Main PID: 3619 (docker)
              CGroup: /system.slice/hyperion-lite.service
                  └─3619 /usr/bin/docker run --rm -v /var/docker/hyperion-lite/elasticsearch:/var/lib/elasticsearch -v /var/docker/hyperion-lite/graphite:/var/lib/graphite/storage/whisper -v /var/docker/hyperion-lite/supervisor:/var/log/supervisor -v /var/docker/hyperion-lite/nginx:/var/log/nginx -p 9990:80 -p 9992:9200 -p 9979:6379 -p 8083:8083 -p 8086:8086 -p 8125:8125/udp -p 2003:2003/tcp --name hyperion-lite nlamirault/hyperion-lite:0.6.0
        $ journalctl -f -u hyperion-lite.service
        fleetctl journal -f hyperion-lite.service
        -- Logs begin at Tue 2014-07-07 23:28:24 UTC. --
        Jul 07 23:28:57 hyperion-lite systemd[1]: Starting hyperion-lite...
        Jul 07 23:28:58 hyperion-lite systemd[1]: Started hyperion-lite.
        Jul 07 23:28:58 hyperion-lite docker[3619]: Unable to find image 'nlamirault/hyperion-lite:0.6.0' locally
        Jul 07 23:28:58 hyperion-lite docker[3619]: Pulling repository nlamirault/hyperion-lite


## Usage

You could use [Hyperion-Lite][] to collect event and logs from hosts.

### Fluentd

Using this file [fluent.conf][] for [Fluentd][] and send logs :

    $ gem install fluentd
    $ gem install fluent-plugin-elasticsearch
    $ fluentd -c fluent.conf

### Heka

Using this file [hekad.toml][] for [Heka][] and send logs :

    $ wget https://github.com/mozilla-services/heka/releases/download/v0.5.2/heka_0.5.2_amd64.deb
    $ dpkg -i heka_0.5.2_amd64.deb
    $ hekad -config=hekad.toml


## Development

* Build the container :

        $ make clean && make

* Setup directories :

        $ sudo mkdir -p /var/docker/hyperion-lite/{elasticsearch,graphite,supervisor,nginx}

* Start the container :

        $ make start

* You could launch unit tests using local installation or VM installation :

        $ pip install -r test-requirements.txt
        $ tox
        $ tox -evm


## Support

Feel free to ask question or make suggestions in our [Issue Tracker][].


## License

Hyperion-Lite is free software: you can redistribute it and/or modify it under the
terms of the GNU General Public License as published by the Free Software
Foundation, either version 3 of the License, or (at your option) any later
version.

Hyperion-Lite is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.  See the GNU General Public License for more details.

See [COPYING][] for the complete license.


## Changelog

A changelog is available [here](ChangeLog.md).


## Contact

Nicolas Lamirault <nicolas.lamirault@gmail.com>



[Hyperion-Lite]: https://github.com/nlamirault/hyperion-lite
[COPYING]: https://github.com/nlamirault/hyperion-lite/blob/master/COPYING
[Issue tracker]: https://github.com/nlamirault/hyperion-lite/issues
[fluent.conf]: https://github.com/nlamirault/hyperion-lite/blob/master/logs/fluent.conf
[hekad.toml]: https://github.com/nlamirault/hyperion-lite/blob/master/logs/hekad.toml

[badge-license]: https://img.shields.io/badge/license-GPL_3-green.svg?style=flat

[Docker]: https://www.docker.io
[Docker documentation]: http://docs.docker.io
[CoreOS]: http://coreos.com
[Etcd]: http://coreos.com/using-coreos/etcd
[Fleet]: http://coreos.com/using-coreos/clustering/
[Nginx]: http://nginx.org
[Elasticsearch]: http://www.elasticsearch.org
[Redis]: http://www.redis.io
[Graphite]: http://graphite.readthedocs.org/en/latest
[Grafana]: http://grafana.org/
[Kibana]: http://www.elasticsearch.org/overview/kibana/
[Carbon]: http://graphite.readthedocs.org/en/latest/carbon-daemons.html
[Statsd]: https://github.com/etsy/statsd/wiki
[ElasticSearchHead]: http://mobz.github.io/elasticsearch-head
[ElasticHQ]: http://www.elastichq.org
[Kopf]: https://github.com/lmenezes/elasticsearch-kopf
[Virtualbox]: https://www.virtualbox.org
[Vagrant]: http://downloads.vagrantup.com
[Fluentd]: http://fluentd.org/
[Heka]: http://hekad.readthedocs.org/en/latest/
[Supervisor]: http://supervisord.org
[sysinfo_influxdb]: https://github.com/novaquark/sysinfo_influxdb
[InfluxDB]: http://influxdb.com
[SystemdD]: http://freedesktop.org/wiki/Software/systemd


ChangeLog.md
Hyperion-Lite ChangeLog
======================

# Version 1.0.0 (2014-12-12)

- Update Grafana dashboard
- Deploy Docker image

# Version 0.6.0 (2014-07-08)

- Split Hyperion and Hyperion-Lite projects
- Update dashboard using Graphite and InfluxDB metrics
- Refactoring unit tests
- Add InfluxDB

# Version 0.5.0 (2014-06-16)

- [#7][] : Update Tox / Python tests configuration to test CoreOS hyperion
- Change Elasticsearch setup using archive and no longer the Debian package
- Update CoreOS setup using Etcd and Fleet
- [#6][] : Add shared volume on Vagrant installation
- [#4][] : Upgrade to Elasticsearch 1.2.1

# Version 0.4.0 (2014-06-05)

- Add CoreOS support
- Add logstash Redis indexer
- `FIX` French timezone on Graphite configuration
- Initialize unit tests
- Configure Logstash to manage nginx logs
- Add Kibana
- Add Logstash
- Add Redis

# Version 0.3.0 (2014-05-23)

- Add Hyperion default dashboard for Grafana
- Update base image to Ubuntu 14.04
- Add Wercker configuration

# Version 0.2.0 (2014-05-21)

- Add Hyperion description file
- Add ElasticSearch plugins
- Configure Nginx
- Add a Statsd client to sent metric to hyperion
- Update Statsd configuration

# Version 0.1.0 (2014-05-16)

- Grafana and Graphite web UI available on the host
- Configure StatsD
- Configure ElasticSearch
- Configure Graphite
- Configure Graphana
- Init project


[#4]: https://github.com/nlamirault/hyperion/issues/4
[#6]: https://github.com/nlamirault/hyperion/issues/6
[#7]: https://github.com/nlamirault/hyperion/issues/7


hyperiontests/test_kibana.py
# Copyright (C) 2014  Nicolas Lamirault <nicolas.lamirault@gmail.com>

# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.


from hyperiontests import hyperion
from hyperiontests import settings


class TestKibana(hyperion.HyperionTestCase):

    def setUp(self):
        super(TestKibana, self).setUp()
        self._host = "http://%s:%s" % (settings.HYPERION_HOST,
                                       settings.HYPERION_WEB)

    def test_can_retrieve_logstash_dashboard(self):
        response = self.http_get("kibana/index.html#/dashboard/file/logstash.json")
        self.assertEqual(200, response.status_code)


hyperiontests/test_grafana.py
# Copyright (C) 2014  Nicolas Lamirault <nicolas.lamirault@gmail.com>

# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.


from hyperiontests import hyperion
from hyperiontests import settings


class TestGrafana(hyperion.HyperionTestCase):

    def setUp(self):
        super(TestGrafana, self).setUp()
        self._host = "http://%s:%s" % (settings.HYPERION_HOST,
                                       settings.HYPERION_WEB)

    def test_can_retrieve_default_dashboard(self):
        response = self.http_get("grafana/#/dashboard/file/default.json")
        self.assertEqual(200, response.status_code)


hyperiontests/test_influxdb.py
# Copyright (C) 2014  Nicolas Lamirault <nicolas.lamirault@gmail.com>

# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.


from hyperiontests import hyperion
from hyperiontests import settings


class TestInfluxDB(hyperion.HyperionTestCase):

    def setUp(self):
        super(TestInfluxDB, self).setUp()
        self._host = "http://%s:%s" % (settings.HYPERION_HOST,
                                       settings.HYPERION_INFLUXDB)

    def _influxdb_request(self, uri):
        response = self.http_get(uri, 'root', 'root')
        self.assertEqual(200, response.status_code)
        self.assertEqual("application/json",
                         response.headers.get("Content-Type"))
        self.assertTrue("InfluxDB v0.7.3" in response.headers.get("X-Influxdb-Version"))
        content = response.json()
        print(content)
        return content

    def test_can_retrieve_influxdb_hyperion_db(self):
        content = self._influxdb_request('db')
        self.assertEqual(1, len(content))
        self.assertEqual('hyperion-lite', content[0]['name'])

    def test_can_retrieve_admin_clusters(self):
        content = self._influxdb_request('cluster_admins')
        self.assertEqual(1, len(content))
        self.assertEqual('root', content[0]['name'])


hyperiontests/hyperion.py
# Copyright (C) 2014  Nicolas Lamirault <nicolas.lamirault@gmail.com>

# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

import logging
import unittest

import docker
import requests
from requests import exceptions


logger = logging.getLogger(__name__)


class HyperionTestCase(unittest.TestCase):

    tmp_imgs = []
    tmp_containers = []

    @classmethod
    def setUpClass(cls):
        pass

    @classmethod
    def tearDownClass(cls):
        pass

    def setUp(self):
        self.client = docker.Client()
        # host = os.getenv('HYPERION_HOST', settings.HYPERION_HOST)
        # self._host = "http://%s:%s" % (host,
        #                                settings.HYPERION_WEB)

    def http_get(self, uri, username=None, password=None):
        """Perform a HTTP GET request.

        :param uri: URL for the new Request object
        """
        try:
            print("GET: %s/%s" % (self._host, uri))
            if username is not None and password is not None:
                return requests.get('%s/%s' % (self._host, uri),
                                    auth=(username, password))
            else:
                return requests.get('%s/%s' % (self._host, uri))
        except exceptions.RequestException as e:
            print ("\033[0;31m%s \033[00m: %s" % ("Error", e))


hyperiontests/settings.py
# Copyright (C) 2014  Nicolas Lamirault <nicolas.lamirault@gmail.com>

# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

HYPERION_HOST = "127.0.0.1"
HYPERION_WEB = 9990
HYPERION_ES = 9992
HYPERION_REDIS = 9979
HYPERION_INFLUXDB = 8086


grafana/influxdb_config.js
/** @scratch /configuration/config.js/1
 * == Configuration
 * config.js is where you will find the core Grafana configuration. This file contains parameter that
 * must be set before Grafana is run for the first time.
 */
define(['settings'],
function (Settings) {
  "use strict";

  return new Settings({

    /**
     * elasticsearch url:
     * For Basic authentication use: http://username:password@domain.com:9200
     */
    //elasticsearch: "http://"+window.location.hostname+":9200",

    /**
     * graphite-web url:
     * For Basic authentication use: http://username:password@domain.com
     * Basic authentication requires special HTTP headers to be configured
     * in nginx or apache for cross origin domain sharing to work (CORS).
     * Check install documentation on github
     */
    //graphiteUrl: "http://"+window.location.hostname+":8080",
      datasources: {
          influx: {
              default: true,
              type: 'influxdb',
              url: 'http://' + window.location.hostname+ ':8086/db/hyperion',
              username: 'root',
              password: 'root',
          }
      },


    /**
     * Multiple graphite servers? Comment out graphiteUrl and replace with
     *
     *  datasources: {
     *    data_center_us: { type: 'graphite',  url: 'http://<graphite_url>',  default: true },
     *    data_center_eu: { type: 'graphite',  url: 'http://<graphite_url>' }
     *  }
     */

    default_route: '/dashboard/file/default.json',

    /**
     * If you experiance problems with zoom, it is probably caused by timezone diff between
     * your browser and the graphite-web application. timezoneOffset setting can be used to have Grafana
     * translate absolute time ranges to the graphite-web timezone.
     * Example:
     *   If TIME_ZONE in graphite-web config file local_settings.py is set to America/New_York, then set
     *   timezoneOffset to "-0500" (for UTC - 5 hours)
     * Example:
     *   If TIME_ZONE is set to UTC, set this to "0000"
     */

    timezoneOffset: null,

    grafana_index: "grafana-dash",

    panel_names: [
      'text',
      'graphite'
    ]
  });
});


client/hyperion_influxdb.py
#!/usr/bin/env python

# Copyright (C) 2014  Nicolas Lamirault <nicolas.lamirault@gmail.com>

# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

import argparse
import socket
import time

import influxdb
import psutil


def format_for_submission(name, values):
    points = [socket.getfqdn()]
    columns = ["host"]
    for field, value in zip(values._fields, values):
        points.append(value)
        columns.append(field)
    return {"points": [points],
            "name": name,
            "columns": columns}


def send_stats(client):
    interval = 1
    while True:
        start_time = time.time()
        data = []
        metrics = {"cpu_percent": psutil.cpu_times_percent,
                   "network_usage": psutil.net_io_counters}
        for name, fn in metrics.iteritems():
            raw_data = fn()
            data.append(format_for_submission(name, raw_data))
        print("Metrics: %s" % data)
        client.write_points(data)
        while (start_time + interval - time.time()) > 0:
            rest_duration = start_time + interval - time.time()
            time.sleep(rest_duration)


def parse_args():
    parser = argparse.ArgumentParser(description='Hyperion InfluxDB client.')
    parser.add_argument('--server',
                        type=str,
                        help=' The influxdb host to connect to, defaults to localhost',
                        default='localhost')
    parser.add_argument('--port',
                        type=int,
                        help='The influxdb port to connect to, defaults to 8086',
                        default=8086)
    parser.add_argument('--dbname',
                        type=str,
                        help='The influxdb database to connect to, defaults to hyperion-lite',
                        default='hyperion-lite')
    parser.add_argument('--username',
                        type=str,
                        help='The influxdb username, defaults to root',
                        default='root')
    parser.add_argument('--password',
                        type=str,
                        help='The influxdb password to connect to, defaults to root',
                        default='root')
    return parser.parse_args()


if __name__ == '__main__':
    args = parse_args()
    client = influxdb.InfluxDBClient(args.server,
                                     args.port,
                                     args.username,
                                     args.password,
                                     args.dbname)
    time.sleep(1)
    send_stats(client)
    # dbusers = client.get_database_users()
    # print("Get list of database users: {0}".format(dbusers))


client/hyperion.sh
#!/bin/bash

# Copyright (C) 2014  Nicolas Lamirault <nicolas.lamirault@gmail.com>

# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

NAME=hyperion-lite
CONTAINER=nlamirault/$NAME
HYPERION_DIR="/var/docker/$NAME"
HYPERION_WEB=9990
HYPERION_ES=9992
HYPERION_REDIS=9979
HYPERION_INFLUXDB_UI=8083
HYPERION_INFLUXDB_API=8086

hyperion_pull() {
    sudo docker pull $CONTAINER
}

hyperion_start() {
    sudo mkdir -p $HYPERION_DIR/{elasticsearch,graphite,supervisor,nginx,redis,influxdb}
    sudo chmod -R 777 $HYPERION_DIR/elasticsearch
    docker run \
           -v $HYPERION_DIR/elasticsearch:/var/lib/elasticsearch \
           -v $HYPERION_DIR/graphite:/var/lib/graphite/storage/whisper \
           -v $HYPERION_DIR/supervisor:/var/log/supervisor \
           -v $HYPERION_DIR/nginx:/var/log/nginx \
           -v $HYPERION_DIR/influxdb:/var/lib/influxdb \
           -p $HYPERION_WEB:80 \
           -p $HYPERION_ES:9200 \
           -p $HYPERION_REDIS:6379 \
           -p $HYPERION_INFLUXDB_UI:8083 \
           -p $HYPERION_INFLUXDB_API:8086 \
           -p 8125:8125/udp -p 2003:2003/tcp \
           -p 9922:22 \
           --name $NAME $CONTAINER
}

hyperion_stop() {
    docker stop $NAME
}

hyperion_clean() {
    docker rm $NAME
}

hyperion_help() {
    echo "Usage: $0 <command>"
    echo "Commands:"
    echo "  pull      :    Pull the Hyperion image from the registry"
    echo "  start     :    Start Hyperion container"
    echo "  stop      :    Stop Hyperion container"
    echo "  clean     :    Remove Hyperion container"
    echo "  help      :    Display this help"
}

if [ $# -eq 0 ]
then
    hyperion_help
    exit 0
fi

case $1 in
    pull)
        hyperion_pull
        ;;
    clean)
        hyperion_clean
        ;;
    start)
        hyperion_start
        ;;
    stop)
        hyperion_stop
        ;;
    *)
        hyperion_help
        ;;
esac
